### PR TITLE
[Collections] Signing (apple, 2): signed collections

### PR DESCRIPTION
**Pull request by yim-lee**
This is part 2 of a series of PRs to support package collection signing on **Apple** platforms. Originally #3242.

Depends on [part 1](#3259)

Modifications:

- Add support for signing with EC and RSA keys
- Add `Signature`, which is similar to JWS
- Introduce `CertificatePolicy` protocol (more on this in part 3)
- `PackageCollectionsSigning` - given a package collection, certificate and its private key, generate a "signed" collection; and reverse that process to validate collection signature.

### CODE REVIEW

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: Sources/PackageCollectionsSigning/Signing/Signing+RSAKey.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import struct Foundation.Data
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+import Security
+#endif
+
+// MARK: - MessageSigner and MessageValidator conformance using the Security framework
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```
Comment (tomerd) [inline Sources/PackageCollectionsSigning/Signing/Signing+RSAKey.swift:19]: maybe we can use `#if canImport(Security)` instead

File: Sources/PackageCollectionsSigning/Signing/Signing+RSAKey.swift [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import struct Foundation.Data
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```
Comment (tomerd) [inline Sources/PackageCollectionsSigning/Signing/Signing+RSAKey.swift:13]: maybe we can use `#if canImport(Security)` instead

File: Sources/PackageCollectionsSigning/Utilities/Base64URL.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    func base64URLEncodedBytes() -> Data {
+        var data = Data(self).base64EncodedData()
+        data.base64URLEscape()
+        return data
+    }
+}
+
+extension Data {
```
Comment (tomerd) [inline Sources/PackageCollectionsSigning/Utilities/Base64URL.swift:42]: nit: this could probably be moved to the Basics module

**Review by tomerd — APPROVED** [review]

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
cc @catmurdoch

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test macOS

**Pull request merged**